include/SH3/system/glvertarray.hpp

```diff
@@ -0,0 +1,250 @@
+         *  @param type - Data Type of data passed in.
+         *  @param size - The size of each component per attribute.
+         *  @param stride - Byte offset between vertex attributes.
+         *  @param offset - Offset to attribute in buffer. Each attrib is then located at offset + stride.
+         */
+        void SetDataLocation(GLuint attrib, DataType type, GLint size, GLsizei stride, GLint offset);
+
+        /**
+         *  Overload the [] operator to get a buffer stored in @ref buffers more easily.
+         *
+         *  @param slot - Index slot where the buffer is stored. Usually in a @ref Attribute.
+         *
+         *  @return buffer_object& reference to the buffer object stored in @ref buffers.
+         */
+        buffer_object& operator[] (GLuint slot)
+        {
+            return buffers[slot];
+        }
+    protected:
+        std::array<buffer_object, numbuffers> buffers; /**< Our list of current VBOs. */
+
+    private:
+
+        /**
+         *  Get offset (stride) into a buffer stored in @ref buffers
+         *
+         *  @param offset - Offset to the first element.
+         *
+         *  @return void* - Pointer to the first element of stride 'offset'.
+         */
+        static constexpr void* buffer_offset(GLint offset){return static_cast<char *>(nullptr) + offset;}
+    };
+
+    /**
+     *  Set vertex shader location of data in VBO for specified attribute.
+     *
+     *  @tparam numbuffers - Number of buffers
+     *  @param attrib - Attribute we want to map data input in our shader (using location = x).
+     *  @param type - Data Type of data passed in.
+     *  @param size - The size of each component per attribute.
+     *  @param stride - Byte offset between vertex attributes.
+     *  @param offset - Offset to attribute in buffer. Each attrib is then located at offset + stride.
+     */
+    template<std::size_t numbuffers>
+    void mutablevao<numbuffers>::SetDataLocation(GLuint attrib, DataType type, GLint size, GLsizei stride, GLint offset)
+    {
+        Bind();
+        glEnableVertexAttribArray(attrib);
+        ASSERT_MSG(attrib < numbuffers, "VBO index (attribute) out of range");
+        buffers[attrib].Bind();
+        glVertexAttribPointer(attrib, size, static_cast<GLenum>(type), GL_FALSE, stride, buffer_offset(offset));
+    }
+
+    /**
+     *  A VAO with indexed vertices.
+     *
+     *  You probably want to use @ref vao.
+     */
+    template<std::size_t numbuffers, std::size_t index_slot>
+    struct indexedmutablevao : public mutablevao<numbuffers>
+    {
+    protected:
+        indexedmutablevao(const buffer_object& vertices): mutablevao<numbuffers>(mutablevao<numbuffers>::buffers[index_slot], numbuffers, vertices) {}
+    };
+
+    /**
+     *  Declares parent type for @ref vao.
+     *
+     *  Either @ref indexedmutablevao or @ref mutablevao (unindexed).
+     *
+     *  @tparam Attribute - Attribute enum (specifies Attribute::MAX to send to @ref mutablevao).
```

Comment:
Also document the usage of `Attribute::INDEX_SLOT`.

I'd just leave a short parameter description for `@tparam` and put the description of it below:
```
@tparam Attribute An attribute enum.

@p Attribute should have a value @c MAX to indicate the number of attributes (slots).
If @p Attribute has a value @c INDEX_SLOT, then the @ref type is an @ref indexedmutablevao.
```